### ISSUE TITLE
[rtl,flash_ctrl] Fix scrambler interface

### ISSUE THREAD

**Issue by matutem** [lowRISC/opentitan#22443]
### Description

The fix for the flash_ctrl scrambler interface is not complete, as we still get some unexpected ecc failures in the flash_ctrl_phy_arb_redun unit test. Specifically, running it with the fix in #22444 with seed 18977237425298456515035329733300663925435727379005790363212818518281951604587 the test shows an ECC error for a read to address 0x08a0 from bank 0 around time 269437 x 100 ps.

From the test waves, the data from the flash primitive and the memory backdoor read is 0xcdf_0523ab48_aa8035de (the top 3 nibbles are both eccs), and the lower 64 bits are never sent to the descrambler, as you can see from https://github.com/lowRISC/opentitan/blob/master/hw/ip/flash_ctrl/rtl/flash_phy_core.sv#L554. 

It can be helpful to add instrumentation in the descramble function at https://github.com/lowRISC/opentitan/blob/master/hw/ip/flash_ctrl/dv/env/flash_otf_item.sv#L173 to show the raw scrambled data, the ecc, and the descrambled data.
![Screenshot from 2024-04-04 23-50-19](https://github.com/lowRISC/opentitan/assets/80294896/139613f0-7963-4e90-ac3a-046c0b2560a4)

**Comment by vogelpi** [comment]
Thanks for filing the issue @matutem ! I've now investigated this a bit and noted the following:
- Your branch `[flash_ctrl_failures](https://github.com/matutem/opentitan/tree/flash_ctrl_failures)` doesn't include the latest fix to the scrambling.
- When rebasing the your branch on current master, I still see a failure with the same seed, but it seems to be slightly different.
- I also see an unexpected ECC error being generated. It's generated for a controller request. To arbitrate between controller and host requests, there is another `prim_arbiter_tree_dup` instance in the design. This one has been there before sharing the scrambling unit and it's the one that the test actually forces.
- The transaction producing the ECC error goes into the scrambling arbitration right after releasing the force signal on the controller / host arbiter. 

I am not yet sure what exactly happens and why the ECC error is produced. It seems to be triggered by the test. Maybe the forcing violates one of the assumptions stated in the design for controller requests (no back-to-back requests, only forwarded if the pipeline is completely empty), or the injected fault / signal force leaves the read pipeline in a weird state. But I think we should fix the RTL here. I'll continue my investigation and report back.

**Comment by matutem** [comment]
I rebased and the failure is still the same on my side. Also, the arbitration fault was injected at time 236438 and ended at 266938, while the ecc error happens at 272438, which is sufficiently far in time. Perhaps I had not updated the PR itself, but it is all square now.

**Comment by matutem** [comment]
@vogelpi, I agree it would be good to improve the test. I would like to understand some points you raise:
- If the arbiters are truly copies used to verify they act in sync then I don't understand that "This is the copy interfacing the read pipeline." in point 1. That seems to imply the other copy doesn't interface the read pipeline, which would mean they are NOT mutually redundant.
- I don't understand something in point 2: "I can see that both the host and the controller requests are de-asserted, but only one of them gets granted.", do you mean are asserted? If they are de-asserted they should not be granted. I tend to check for protocol violations like this using DV SVAs, and that would be good to add.
- Apparently the ECC errors can happen some time after the fault insertion: the case I was examining has the ECC error a good number of cycles after the fault injection stops. This could impact your point 3, since a fetch for the NMI could be the one that gets the ecc error, and getting faults when handling a fault is a potential disaster.

I had trouble causing a clean end for the test, but I can give it another try.

**Comment by matutem** [comment]
By the way, apparently we have two functionally different arbiters, each with a redundant one. I imagine we should introduce faults in each pair, so we may be missing a test. Were you concerned about this in your comments above?

**Comment by vogelpi** [comment]
> @vogelpi, I agree it would be good to improve the test. I would like to understand some points you raise:
> 
>     * If the arbiters are truly copies used to verify they act in sync then I don't understand that "This is the copy interfacing the read pipeline." in point 1. That seems to imply the other copy doesn't interface the read pipeline, which would mean they are NOT mutually redundant.

If you study the RTL of `prim_arbiter_tree_dup`, you'll see that the output of the last arbiter (Index 1 if two copies are used) drives the output of the primitive. The other arbiters just feed into the error detection. The inputs to all arbiters are the same though.

>     * I don't understand something in point 2: "I can see that both the host and the controller requests are de-asserted, but only one of them gets granted.", do you mean are asserted? If they are de-asserted they should not be granted. I tend to check for protocol violations like this using DV SVAs, and that would be good to add.

Yes, they are both de-asserted but only one gets a grant. It's a protocol violation.

>     * Apparently the ECC errors can happen some time after the fault insertion: the case I was examining has the ECC error a good number of cycles (more than 3 microseconds) after the fault injection stops. This could impact your point 3, since a fetch for the NMI could be the one that gets the ecc error, and getting faults when handling a fault is a potential disaster.
> 
>     * The test actually waits between 600 and 900 micros before applying a reset: https://github.com/lowRISC/opentitan/blob/master/hw/ip/flash_ctrl/dv/env/seq_lib/flash_ctrl_phy_arb_redun_vseq.sv#L29.

My understanding based on the discussion on TUE is that:
1. We don't expect flash_ctrl to continue operating after inserting the fault. It's nice to have but we shouldn't fail the test if it doesn't.
2. We should ignore other errors raised after inserting the fault. What we care about is the fault and signal the fatal alert. Everything else is opportunistic and we don't need to give guarantees. Right now, we already ignore another error, IIRC it's called mp_err or something. There is a comment in the test. Once you'll also glitch copy 1, the test actually inserts errors that propagate downstream. It's unrealistic to fix the design to keep operating in that scenario.

> By the way, apparently we have two functionally different arbiters, each with a redundant one. I imagine we should introduce faults in each pair, so we may be missing a test. Were you concerned about this in your comments above?

To be precise, we have 3 arbiters:
- One inside flash_phy_core arbiting requests coming in from the host and the controller interfaces. This is the only one glitched by this test currently.
- Two inside flash_phy_scramble. Their error signal connects to the same err bit and alert. But they are not tested currently and I agree we should test them as well.

**Comment by matutem** [comment]
We determined the ECC failures are not related to the arbiter fault injection, so there is serious bug in flash_ctrl.

One reason they may not have been found earlier at the unit level is because the legacy scoreboard does not enable ECC, so this would only be detected in tests running with the otf scoreboard.

**Comment by vogelpi** [comment]
For the record: after some more debugging we've found an RTL in the read pipeline that was present in both versions of flash_ctrl (w/ and w/o shared scrambling logic). In a nutshell, the design always assumed the mask computation to return faster than the Flash read. If for whatever reason, the Flash read was faster than the mask computation and the Flash operation was an erase, the mask storage FIFO got popped (to drop the computed mask) before the mask was actually written. At this point, the mask and the data storage FIFOs got out of sync and any read operation to scrambled locations would return garbage data and ECC errors. This got now fixed with https://github.com/lowRISC/opentitan/pull/22571.

Additional DV tasks for M4 (or potentially V3 closure - TBC) are captured in https://github.com/lowRISC/opentitan/issues/22559 .
